From 160210e335474dd72124e85461ffec6db1b75329 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Petr=20P=C3=ADsa=C5=99?= <ppisar@redhat.com>
Date: Fri, 18 Jan 2019 13:25:06 +0100
Subject: [PATCH] Respect width and height unsigness in arithmetics
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

XParseGeometry() and various other X11 functions expect width and
height as unsigned integers. This was fixes in 1.3. However, the
variables are used in expressions that coerces into signed variables
and we could experience an arithmetics overflow. The overflow is
prevented by limiting width's and height's value. Also comparision
between signed and unsigned is undefined.

Possible multiplication overflows in xrealloc() arguments are not
addressed.

Signed-off-by: Petr Písař <ppisar@redhat.com>
---
 root-tail.c | 35 +++++++++++++++++++++++++++--------
 1 file changed, 27 insertions(+), 8 deletions(-)

diff --git a/root-tail.c b/root-tail.c
index a85dcb2..c41b2c7 100644
--- a/root-tail.c
+++ b/root-tail.c
@@ -38,6 +38,7 @@
 #include <locale.h>
 #include <ctype.h>
 #include <stdarg.h>
+#include <limits.h>
 #include <X11/Xlib.h>
 #include <X11/Xatom.h>
 #include <X11/Xutil.h>
@@ -118,7 +119,7 @@ static int continuation_length;
 /* HACK - ideally listlen will start at however many '~'s will fit on
  * the screen */
 static unsigned int width = STD_WIDTH, height = STD_HEIGHT;
-static int listlen = 50;
+static unsigned int listlen = 50;
 static int win_x = LOC_X, win_y = LOC_Y;
 static int effect_x_space, effect_y_space; /* how much space does shading / outlining take up */
 static int effect_x_offset, effect_y_offset; /* and how does it offset the usable space */
@@ -369,7 +370,7 @@ InitWindow (void)
 	{
 	  fprintf(stderr, "\n  the display isn't tall enough to display a single line in font '%s'\n",
 		  e->fontname);
-	  fprintf(stderr, "\n  the geometry in use is %d pixels tall\n", height);
+	  fprintf(stderr, "\n  the geometry in use is %u pixels tall\n", height);
 	  fprintf(stderr, "\n  font '%s' is %d pixels tall\n", e->fontname, e->font_height);
 	  if (effect_y_space)
 	    fprintf(stderr, "\n  the shade or outline options need an extra %d pixel%s of vertical space\n",
@@ -445,8 +446,8 @@ draw_text (Display *disp, Window root, GC WinGC, int x, int y, struct line_node
 static void
 refresh (int miny, int maxy, int clear, int refresh_all)
 {
-  int lin = 0;
-  int space = height;
+  unsigned int lin = 0;
+  int space = (int)height;
   int offset;
   unsigned long black_color = GetColor ("black");
   struct line_node *line;
@@ -470,7 +471,7 @@ refresh (int miny, int maxy, int clear, int refresh_all)
 
       if (opt_noflicker && lin >= listlen)
         {
-          int i = listlen;
+          unsigned int i = listlen;
           listlen *= 1.5;
           display = xrealloc(display, listlen * sizeof(struct displaymatrix));
           for (; i < listlen; i++)
@@ -1112,7 +1113,7 @@ append_to_existing_line (char *str, struct logfile_entry *log)
 static void
 main_loop (void)
 {
-  int lin;
+  unsigned int lin;
   time_t lastreload;
   Region region = XCreateRegion ();
   XEvent xev;
@@ -1325,8 +1326,26 @@ main (int argc, char *argv[])
           else if (!strcmp (arg, "-V"))
             display_version ();
           else if (!strcmp (arg, "-g") || !strcmp (arg, "-geometry"))
-            geom_mask =
-              XParseGeometry (argv[++i], &win_x, &win_y, &width, &height);
+            {
+              geom_mask =
+                XParseGeometry (argv[++i], &win_x, &win_y, &width, &height);
+              /* Width and height are used in various expressions assigned to
+               * signed or unsigned integers. The assignments could overflow. */
+              if (width < 1 || width > INT_MAX - MARGIN_OF_ERROR || width > UINT_MAX)
+                {
+                  fprintf (stderr,
+                      "The geometry width is too small or too large: '%u'.\n",
+                      width);
+                  exit (1);
+                }
+              if (height < 1 || height > UINT_MAX)
+                {
+                  fprintf (stderr,
+                      "The geometry height is too small or too large: '%u'.\n",
+                      height);
+                  exit (1);
+                }
+            }
           else if (!strcmp (arg, "-display"))
             dispname = argv[++i];
           else if (!strcmp (arg, "-cont"))
-- 
2.21.0

